lib/Dialect/FIRRTL/Transforms/HoistPassthrough.cpp

```diff
@@ -0,0 +1,571 @@
+    return dyn_cast<OpResult>(getDest());
+  }
+
+  /// Helper to obtain argument/result number of destination.
+  /// Must be block arg or op result.
+  static size_t getIndex(Value v) {
+    if (auto arg = dyn_cast<BlockArgument>(v))
+      return arg.getArgNumber();
+    auto result = dyn_cast<OpResult>(v);
+    assert(result);
+    return result.getResultNumber();
+  }
+};
+
+/// Driver implementation for probes.
+struct RefDriver : public Driver {
+  using Driver::Driver;
+
+  static bool classof(const Driver *t) { return isa<RefValue>(t->getDest()); }
+
+  static RefDriver get(Value v);
+
+  Value remat(PortMappingFn mapPortFn, ImplicitLocOpBuilder &builder);
+};
+static_assert(sizeof(RefDriver) == sizeof(Driver),
+              "passed by value, no slicing");
+
+// Driver implementation for HW signals.
+// Split out because has more complexity re:safety + updating.
+// And can't walk through temporaries in same way.
+struct HWDriver : public Driver {
+  using Driver::Driver;
+
+  static bool classof(const Driver *t) { return !isa<RefValue>(t->getDest()); }
+
+  static HWDriver get(Value v);
+
+  Value remat(PortMappingFn mapPortFn, ImplicitLocOpBuilder &builder);
+};
+static_assert(sizeof(HWDriver) == sizeof(Driver),
+              "passed by value, no slicing");
+
+/// Print driver information.
+template <typename T>
+static inline T &operator<<(T &os, Driver &d) {
+  if (!d)
+    return os << "(null)";
+  return os << d.getDest() << " <-- " << d.drivingConnect << " <-- "
+            << d.source.getValue() << "@" << d.source.getFieldID();
+}
+
+} // end anonymous namespace
+
+//===----------------------------------------------------------------------===//
+// Driver implementation.
+//===----------------------------------------------------------------------===//
+
+Driver Driver::get(Value v) {
+  if (auto refDriver = RefDriver::get(v))
+    return refDriver;
+  if (auto hwDriver = HWDriver::get(v))
+    return hwDriver;
+  return {};
+}
+
+Value Driver::remat(PortMappingFn mapPortFn, ImplicitLocOpBuilder &builder) {
+  return TypeSwitch<Driver *, Value>(this)
+      .Case<RefDriver, HWDriver>(
+          [&](auto *d) { return d->remat(mapPortFn, builder); })
+      .Default({});
+}
+
+void Driver::finalize(ImplicitLocOpBuilder &builder) {
+  auto immSource = drivingConnect.getSrc();
+  auto dest = getDest();
+  assert(immSource.getType() == dest.getType() &&
+         "final connect must be strict");
+  if (dest.hasOneUse()) {
+    // Only use is the connect, just drop it.
+    drivingConnect.erase();
+  } else if (isa<BlockArgument>(immSource)) {
+    // Block argument dominates all, so drop connect and RAUW to it.
+    drivingConnect.erase();
+    dest.replaceAllUsesWith(immSource);
+  } else {
+    // Insert wire temporary.
+    // For hoisting use-case could also remat using cached indexing inside the
+    // module, but wires keep this simple.
+    auto temp = builder.create<WireOp>(immSource.getType());
```

Comment:
Is there an example for this case in the test ? Does this leave the wire uninitialized ?

Reply:
No, this is for the case where the wire is needed internally because the port has users (not just the driver).

And yep there's a test, see `NeedsWire`.